### ISSUE TITLE
[HOLD for payment 2023-07-05] [$2000] iOS - Keyboard does not hide when switching between attachments with focused password input field

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#19156]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the App
2. Login with any account
3. Send several attachments including PDF-protected file
4. Click to preview on PDF-protected file -> Press Enter a password
5. Switching to another attachment

## Expected Result:
Keyboard should be hidden

## Actual Result:
Keyboard does not hide 

## Workaround:
unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [X] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.15-6
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856 
**Notes/Photos/Videos:** Any additional supporting documentation 

https://github.com/Expensify/App/assets/43995119/6f87bbd7-de1d-4df5-8197-a060b73f80c0

**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team 
**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014d32d06c969c5ab1</li>
        <li>Upwork Job ID: 1661854565030178816</li>
        <li>Last Price Increase: 2023-06-21</li>
    </ul>
</details>

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Keyboard does not hide when switching between attachments with focused password input field.
### What is the root cause of that problem?
there is no Keyboard.dismiss(); when next/prev button press
### What changes do you think we should make in order to solve the problem?
call Keyboard.dismiss(); when next/prev button onPress

### What alternative solutions did you explore? (Optional)

**Comment by sonialiap** [comment]
I'm unable to reproduce on the browserstack emulator on iPhone 13 v16.3

<img width="381" alt="image" src="https://github.com/Expensify/App/assets/17131195/d844f41a-77c1-4e5f-9764-9c1b784d5389">

**Comment by sonialiap** [comment]
@ahmedGaber93 since you proposed a solution, are you able to reproduce this issue?

**Comment by sonialiap** [comment]
Thanks! Adding the external label

**Comment by ahmedyasin21** [comment]
Myself Ahmed Yasin, a certified React Native Developer.

Since you're using keyboard.dismiss() . it'll surely work if you place it on a right place and in a right way. Once you flow the documentation right. It'll be solve. I have faced this issue while developing one of my app. And I resolved and saved the solution in may notes. I do this with every single error I faced.

I'm right for this job. So , hoping of response.

**Comment by ahmedyasin21** [comment]
Contributor details
Your Expensify account email: ahmedyasin1947@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/mahmedyasin


Myself Ahmed Yasin, a certified React Native Developer.

Since you're using keyboard.dismiss() . it'll surely work if you place it on a right place and in a right way. Once you follow the documentation right. It'll be solve. I have faced this issue while developing one of my app. And I resolved and saved the solution in may notes. I do this with every single error I faced.

I'm confident and feel myself a best contributor for this job. So , hoping for your response.

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Keyboard does not hide when switching between attachments with focused PDF password input field

### What is the root cause of that problem?
When navigating by swipe left, right, iOS will automatically dismiss any open keyboard.

But not when navigating between attachments by clicking on arrow left, right. We're not manually dismissing the Keyboard in that case, so the Keyboard remains open.

### What changes do you think we should make in order to solve the problem?

There're 2 changes we need to make:
1. Add `Keyboard.dismiss()` in `onPress` when we click on the arrow left, right to navigate here https://github.com/Expensify/App/blob/ec9bc04108992ff400b3fb86d8a31091d727310e/src/components/AttachmentCarousel/index.js#L307 and here https://github.com/Expensify/App/blob/ec9bc04108992ff400b3fb86d8a31091d727310e/src/components/AttachmentCarousel/index.js#L326. We can optionally check if the keyboard is open and only dismiss if true, if keyboard is not open there's no need to dismiss it.

2. We need to make sure when navigate back to the PDF password from, the input will be focused again, this is because the password field is the only input in the page, so it's important to bring up the Keyboard again if we navigate back to it, for the best UX.
- Add a new `activeSource` state to the `AttachmentCarousel` to store the current active attachment
- Add `isFocused={item.source === this.state.activeSource}` to `AttachmentView` when rendering it here https://github.com/Expensify/App/blob/ec9bc04108992ff400b3fb86d8a31091d727310e/src/components/AttachmentCarousel/index.js#LL264C6-L264C6
- Pass the isFocused through the `PDFView`, `PDFPasswordForm` as prop, or we can use context to do this if we want to
- In https://github.com/Expensify/App/blob/87bb08b451ec116e2bf9976a8f4b00f0a9435d90/src/components/PDFView/PDFPasswordForm.js#L43, add logic to focus the input once the PDF Form is focused 
```
if (!prevProps.isFocused && this.props.isFocused && this.textInput) {
    this.textInput.focus();
}
 ``` 

### What alternative solutions did you explore? (Optional)
NA

**Comment by MariaHCD** [comment]
Not overdue, waiting on @aimane-chnaif to review the proposals above

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by MariaHCD** [comment]
> I like this improvement. Can this be fixed here or out of scope?

@aimane-chnaif I like that improvement too! I think it's within the scope of this issue

**Comment by MariaHCD** [comment]
> If you feel this requires more efforts as it fixes multiple bugs related to password protected pdf, we can increase bounty.

Agreed with @aimane-chnaif. I think it would be better to fix the bugs associated with password protected PDF attachments and the keyboard as part of this issue. We can definitely increase the bounty if there is more effort and complexity involved. What do you think, @sonialiap?

So currently, the scope of the solution includes:
1. When the user navigates away from the PDF password form by clicking on the arrow buttons, dismiss the keyboard
2. When the user clicks on the `eye` icon and navigates away from the form, dismiss the keyboard
3. When the navigates back to the PDF password form, focus the input and bring up the keyboard again

Is there anything I'm missing?

**Comment by ahmedGaber93** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/19156#issuecomment-1552218437) to cover all 4 point [here](https://github.com/Expensify/App/issues/19156#issuecomment-1573359277)

**Comment by MariaHCD** [comment]
Not overdue, @aimane-chnaif will evaluate the proposals we have already.

**Comment by ahmedGaber93** [comment]
@aimane-chnaif 
> [mWeb] onScrollBeginDrag doesn't work
This is known issue and that's why OptionsList is using platform-specific codes to dismiss keyboard when start dragging.

That's true, we can use [keyboarddismissmode](https://reactnative.dev/docs/scrollview#keyboarddismissmode) like OptionsList

```javascript
// onScrollBeginDrag={Keyboard.dismiss}
keyboardDismissMode="on-drag" // this will support all platform
```



> [iOS] auto-focus doesn't work when swipe between protected PDFs quickly

This because when you swipe quickly, swipe not complete to view the whole PDF.
We can fix it by reduce `itemVisiblePercentThreshold: 95` to 75
https://github.com/Expensify/App/blob/035783a7984e0214677b5364d9351c968887bf58/src/components/AttachmentCarousel/index.js#L53-L57

**Comment by ahmedGaber93** [comment]
@aimane-chnaif 
> [iOS] keyboard glitch

This because keyboard not dismiss immediately when drag start, but after replace `onScrollBeginDrag={Keyboard.dismiss}` with `keyboardDismissMode="on-drag"`, keyboard will dismiss immediately when swipe.

**Comment by ahmedGaber93** [comment]
@aimane-chnaif 
Yes, I am doing that, I explain the both root cause and the solution.
`keyboardDismissMode="on-drag"` will fix the glitch and support [mWeb]
reduce `itemVisiblePercentThreshold: 95` to 75 will fix quick swipe.
And the other is out of scope, as you explain.
And also I found re focus when navigates back to the PDF password form not work in ios safari, so I think we need increase bounty.

**Comment by ahmedGaber93** [comment]
@aimane-chnaif what do you think of the fixes [here](https://github.com/Expensify/App/issues/19156#issuecomment-1586338331)?

**Comment by ahmedGaber93** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/19156#issuecomment-1552218437)

**Comment by sonialiap** [comment]
@aimane-chnaif proposals have been updated, could you please review them?

**Comment by sonialiap** [comment]
@aimane-chnaif if the scope has increased, we can consider increasing the bounty 👍 Do you have a proposal for how much you think the scope has increased and to what we should increase the bounty?

**Comment by MariaHCD** [comment]
That sounds fair to me 👍🏼

**Comment by ahmedGaber93** [comment]
@aimane-chnaif [my updated branch](https://github.com/ahmedGaber93/App/tree/issue-19156).

**Comment by sonialiap** [comment]
The new amount sounds good to me as well, updating the title and upwork bounty

**Comment by MariaHCD** [comment]
@dukenv0307's solution looks good to me too.

**Comment by MariaHCD** [comment]
Oops, this was labelled internal. Resetting back to external.
